
---
title: "Comparison of cell type-specific gene expression signature"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
  
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(width = 80)

dim1 <- function(x){if(is.vector(x)){length(x)} else{dim(x)}}
```

Compare cell type-specific gene expression signatures generated from snRNA-seq measured by **SMART-seq** vs. **10x 3'v3**.

Both datasets measured gene expression of brain cell types from middle temporal gyrus (MTG). The SMART-seq data were published earlier [@hodge2019conserved]. The 10x data were downloaded from [cellxgene](https://cellxgene.cziscience.com/collections/1ca90a2d-2943-483d-b678-b809bf464c30) collection of SEA-AD (Seattle Alzheimerâ€™s Disease Brain Cell Atlas) data. 

The SMART-seq were processed by codes saved at our GitHub repo  [scRNAseq_pipelines](https://github.com/Sun-lab/scRNAseq_pipelines/). We  used different clustering methods to examine whether the clustering results are consistent with cell type annotation, and choose those cells whose cluster membership is consistent with cell type annotation to generate gene expression signature. 

The 10x data were processed by code `step1_SEA_AD.py` in this folder. We simply calculated two cell type-specific gene expression measures for each gene. One is the summation of UMI counts across all the cells of one cell type, and the other is mean value of read-depth normalized UMIs. 

# Setup

## Load R libraries

* `git_dir` is the local dir for the Github repo `scRNAseq_pipelines`.
* `tenx_dir` is the local dir where saved the 10x gene expressio data generated by `step1_SEA_AD.py`. 


```{r load_libraries, warning = FALSE, message = FALSE}
library(data.table)
library(stringr)
library(org.Hs.eg.db)
library(ICeDT)

library(ggplot2)
library(ggpubr)
library(viridis)
library(ggpointdensity)
library(RColorBrewer)

theme_set(theme_classic())

git_dir  = "~/research/GitHub/scRNAseq_pipelines"
tenx_dir = "~/research/data/SEA-AD"

dim1 <- function(x){if(is.vector(x)) length(x) else dim(x)}
is.uique <- function(x){ length(x) == length(unique(x))}
```

# Read in gene expression signature matrix

## Read in SMART-seq data

The data were generated by codes in our Github repo `scRNAseq_pipelines`. Need to read the expression of all the genes in order to calculate TPM.

```{r fig.width=4, fig.height=4}
rds = readRDS(file.path(git_dir, "_brain_cell_type/step3_DE_output", 
                        "all_genes_MTG.rds"))
lapply(rds, dim1)
sm2_data = rds$SIG
sm2_anno = rds$anno
rm(rds)

rds = readRDS(file.path(git_dir, "_brain_cell_type/step3_DE_output", 
                        "signature_MTG.rds"))
lapply(rds, dim1)
sm2_markers = rownames(rds$SIG)
rm(rds)

cts = readRDS(file.path(git_dir, "MTG/ct_matrix_human_MTG.rds"))
dim(cts)
cts[1:2,] 

cts = data.frame(cts)
cts$gene = rownames(cts)

sm2_df = data.frame(sm2_data)
sm2_df$gene = rownames(sm2_data)
sm2_df = merge(sm2_df, cts, by="gene", suffixes=c("", "_count"))
dim(sm2_df)
sm2_df[1:2,]
cor(log10(sm2_df$Exc_count), log10(sm2_df$Exc))

smrt_df = cts[,c(8,1:7)]
dim(smrt_df)
smrt_df[1:2,] 
```

## Read in 10x data

```{r fig.width=4, fig.height=4}
files = list.files(tenx_dir, pattern="\\.csv")
length(files)

inh_cts = c("Lamp5", "Pvalb", "Sst", "Vip")
exc_cts = c("L2_3_IT", "L4_IT", "L5_IT", "L6_IT")

cell_types = gsub("_genes_info.csv", "", files, fixed = TRUE)
cell_types

dat_10x = list()

for(i in 1:length(files)){
  f1 = files[i]
  ct = cell_types[i]
  if(ct %in% inh_cts){ ct = paste0("Inh_", ct) }
  if(ct %in% exc_cts){ ct = paste0("Exc_", ct) }
  
  dat_10x[[ct]] = fread(file.path(tenx_dir, f1))
  
  if(i == 1){
    tenx_df = dat_10x[[ct]][,c("gene_ids", "feature_name")]
  }else{
    stopifnot(all(tenx_df$gene_ids == dat_10x[[ct]]$gene_ids))
  }
  tenx_df[[ct]] = dat_10x[[ct]]$total_counts
}

dim(tenx_df)
tenx_df[1:2,]

L_23 = dat_10x$Exc_L2_3_IT
dim(L_23)
L_23[1:2,]

L_23 = L_23[which(L_23$pct_dropout_by_counts <= 98),]
dim(L_23)

ggplot(L_23, aes(x=log10(total_counts), 
                 y=log10(mean_rd_normalized_count))) +
  geom_point(alpha=0.5, color="orange")

ggplot(L_23, aes(x=log10(mean_rd_normalized_count), 
                 y=log10(var_rd_normalized_count))) +
  geom_point(alpha=0.5, color="orange")
```

## Read in gene annoation

```{r}
gtf_fn = file.path(git_dir, "MTG/GTF2/gene_info_v26.rds")
gtf = readRDS(gtf_fn)
gtf[1:2,]

exon_fn = file.path(git_dir, 
                    "MTG/GTF2/exon_by_genes_gencode_v26.rds")
exon = readRDS(exon_fn)

dim(gtf)
length(exon)

stopifnot(setequal(names(exon), gtf$ENSG))

gtf = gtf[match(names(exon), gtf$ENSG),]
gene_len = sum(BiocGenerics::width(GenomicRanges::reduce(exon)))
gtf$gene_length = as.numeric(gene_len)
rm(exon)

gtf = gtf[which(gtf$Chr %in% paste0("chr",1:22)),]
gtf$ChrNum = as.integer(gsub("chr","",gtf$Chr))
gtf = gtf[order(gtf$ChrNum, gtf$Start),]

gtf$ENSG_version = gtf$ENSG
gtf$ENSG = gsub("\\..*","", gtf$ENSG_version)

dim(gtf)
gtf[1:2,]
```
	
## Check the relation between gene length and expression

### SMART-seq data

```{r fig.width=9, fig.height=6, warning=FALSE}
gene_anno = as.data.frame(sm2_anno[,c("gene", "chromosome", 
                                      "entrez_id")])
dim(gene_anno)
gene_anno[1:2,]
length(unique(gene_anno$gene))

map1 = mapIds(org.Hs.eg.db, keys=as.character(gene_anno$entrez_id), 
              column='ENSEMBL', keytype='ENTREZID')
map1[1:5]
stopifnot(names(map1) == gene_anno$entrez_id)
gene_anno$ENSG = as.character(map1)
table(is.na(gene_anno$ENSG))

gene_anno = gene_anno[which(!is.na(gene_anno$ENSG)),]
t1 = table(gene_anno$ENSG)
table(t1)
t1[t1 > 2]
gene_anno[gene_anno$ENSG == names(t1[t1 > 2])[1],]

non_uniq_ENSG = names(t1)[t1 > 1]

gene_anno = gene_anno[-which(gene_anno$ENSG %in% non_uniq_ENSG),]
dim(gene_anno)
gene_anno[1:2,]

table(gene_anno$ENSG %in% gtf$ENSG)
gene_anno = merge(gene_anno, gtf, by="ENSG")
gene_anno = gene_anno[,c("ENSG", "gene", "Chr", "Start", "End", "gene_length")]
dim(gene_anno)
gene_anno[1:2,]

smrt_df = merge(smrt_df, gene_anno, by="gene")
dim(smrt_df)
smrt_df[1:2,]

cts_smrt = c("Astro", "Exc", "Inh", "Micro", "Oligo", "OPC")
cr1 = cor(smrt_df$gene_length, smrt_df[,cts_smrt], method="spearman")
cr2 = cor(smrt_df$gene_length, smrt_df[,cts_smrt]/smrt_df$gene_length, 
          method="spearman")

glist = list()

for(ct_i in cts_smrt){
  glist[[ct_i]] = ggplot(smrt_df, 
                    aes_string(x="log10(gene_length)", 
                               y=sprintf("log10(%s)", ct_i))) + 
    geom_pointdensity() + scale_color_viridis() + 
    ggtitle(sprintf("%s, Sp. corr: %.2f/%.2f", ct_i, 
                    cr1[,ct_i], cr2[,ct_i])) + 
    theme(legend.position = "none")
}

ggarrange(plotlist=glist, nrow = 2, ncol = 3)

```

### 10x data

```{r fig.width=9, fig.height=12, warning=FALSE}
dim(tenx_df)
cols2kp = c("ENSG", "Chr", "Start", "End", "gene_length")
tenx_df = merge(tenx_df, gtf[,cols2kp], by.x="gene_ids", by.y="ENSG")
names(tenx_df)[1:2] = c("ENSG", "gene")
names(tenx_df)[which(names(tenx_df) == "Micro-PVM")] = "Micro"

dim(tenx_df)
tenx_df[1:2,]

cts_10x = names(tenx_df)[3:14]
cts_10x

glist = list()

for(i in 1:length(cts_10x)){
  ct1 = cts_10x[i]
  cr1 = cor(tenx_df$gene_length, tenx_df[[ct1]], method="spearman")
  cr2 = cor(tenx_df$gene_length, tenx_df[[ct1]]/tenx_df$gene_length, 
            method="spearman")

  glist[[i]] = ggplot(tenx_df, aes_string(x="log10(gene_length)", 
                 y=sprintf("log10(%s)", ct1))) + 
    geom_pointdensity() + scale_color_viridis() + 
    ggtitle(sprintf("%s, Sp. corr: %.2f/%.2f", ct_i, cr1, cr2)) + 
    theme(legend.position = "none")
}

ggarrange(plotlist=glist, nrow = 4, ncol = 3)

```

# Check correlation between two datasests

## Using refined excitatory and inhibitory neuron types

```{r fig.width=5, fig.height=3}
theme1 = theme(axis.text.x = element_text(angle = 60, vjust = 1, 
                                   size = 11, hjust = 1),
               axis.text.y = element_text(size = 11))
stopifnot(is.uique(smrt_df$ENSG))
stopifnot(is.uique(tenx_df$ENSG))

dat_combined = merge(smrt_df, tenx_df, by="ENSG", 
                     suffixes = c("","_10x"))
dim(dat_combined)
dat_combined[1:2,]

table(dat_combined$gene == dat_combined$gene_10x)

cts_10x_adj = cts_10x
ww1 = which(cts_10x %in% cts_smrt)
cts_10x_adj[ww1] = paste0(cts_10x[ww1], "_10x")

cr_mat = cor(dat_combined[,cts_smrt], dat_combined[,cts_10x_adj], 
             method="spearman")
round(cr_mat, 2)

length(sm2_markers)
sm2_markers[1:4]
table(sm2_markers %in% dat_combined$gene)

dat_markers = dat_combined[which(dat_combined$gene %in% sm2_markers),]
dim(dat_markers)
dat_markers[1:2,]

cr_mat = cor(dat_markers[,cts_smrt], dat_markers[,cts_10x_adj], 
             method="spearman")
colnames(cr_mat) = gsub("_10x", "", colnames(cr_mat))

cr_mat = cr_mat[,order(colnames(cr_mat))]
round(cr_mat, 2)

melted_cr_mat = reshape2::melt(cr_mat, value.name = "corr", 
                               varnames=c("cell type SMART-seq",
                                          "cell type 10x"))
dim(melted_cr_mat)
melted_cr_mat[1:2,]

ggplot(data = melted_cr_mat, 
       aes(x=`cell type 10x`, y=`cell type SMART-seq`, fill=corr)) + 
  geom_tile() + theme_classic() + 
  scale_fill_gradient(low = "white", high = "red") + theme1
```

## Using collapsed excitatory and inhibitory neuron types

```{r fig.width=4, fig.height=3}
exc_cts = c("L2_3_IT", "L4_IT", "L5_IT", "L6_IT")
inh_cts = c("Lamp5", "Pvalb", "Sst", "Vip")

dat_combined[["Exc_10x"]] = rowSums(dat_combined[,paste0("Exc_", exc_cts)])
dat_combined[["Inh_10x"]] = rowSums(dat_combined[,paste0("Inh_", inh_cts)])

cts_10x_combined = paste0(cts_smrt, "_10x")

dat_markers = dat_combined[which(dat_combined$gene %in% sm2_markers),]
dim(dat_markers)
dat_markers[1:2,]

cr_mat = cor(dat_markers[,cts_smrt], dat_markers[,cts_10x_combined], 
             method="spearman")
colnames(cr_mat) = gsub("_10x", "", colnames(cr_mat))

cr_mat = cr_mat[,order(colnames(cr_mat))]
round(cr_mat, 2)

melted_cr_mat = reshape2::melt(cr_mat, value.name = "corr", 
                               varnames=c("cell type SMART-seq",
                                          "cell type 10x"))
dim(melted_cr_mat)
melted_cr_mat[1:2,]

ggplot(data = melted_cr_mat, 
       aes(x=`cell type 10x`, y=`cell type SMART-seq`, fill=corr)) + 
  geom_tile() + theme_classic() + 
  scale_fill_gradient(low = "white", high = "red") + theme1
```


## Check a few scatter plot
```{r fig.width=4, fig.height=4}
for(ct1 in cts_smrt){
  g1 = ggplot(dat_combined, aes_string(x=sprintf("log10(%s)", ct1), 
                   y=sprintf("log10(%s_10x)", ct1))) +
    geom_point(alpha=0.5, color="orange") + 
    geom_abline(intercept = 0, slope=1, col="red")
  print(g1)
}

```

# Deconvolution

## Read in bulk RNA-seq data
```{r}
bulk = readRDS("../../CSeQTL/s7_gtexBrain/counts/trec_hap1_hap2_QCres.rds")
sapply(bulk, dim1)
bulk = bulk$trec

dim(bulk)
bulk[1:2,1:3]
rownames(bulk) = gsub("\\..*", "", rownames(bulk))

dim(dat_combined)
table(dat_combined$ENSG %in% rownames(bulk))
bulk = bulk[match(dat_combined$ENSG, rownames(bulk)),]
dim(bulk)
```


## Calculate TPM using all available genes

```{r fig.width=9, fig.height=6}
num_genes_tpm = nrow(bulk); num_genes_tpm

get_tpm <- function(dat_bulk, gene_length){
  dat_tpm 	= apply(dat_bulk,2,function(xx) xx / gene_length)
  dat_tpm 	= 1e6 * apply(dat_tpm,2,function(xx) xx / sum(xx))
  dat_tpm
}

bulk_tpm 	= get_tpm(bulk, dat_combined$gene_length)

dim(bulk)
dim(bulk_tpm)
bulk[1:2,1:3]
bulk_tpm[1:2,1:3]

dim(dat_combined)
dat_combined[1:2,]

cell_types = c("Astro", "Exc", "Inh", "Micro", "Oligo", "OPC")
cell_types_10x = paste0(cell_types, "_10x")
  
sig_sm2 = dat_combined[,cell_types]
sig_10x = dat_combined[,cell_types_10x]

dim(sig_sm2)
sig_sm2[1:2,]

dim(sig_10x)
sig_10x[1:2,]

rownames(sig_sm2) = dat_combined$ENSG
rownames(sig_10x) = dat_combined$ENSG

length(sm2_markers)
sm2_markers[1:5]

table(sm2_markers %in% dat_combined$gene)
markers = dat_combined$ENSG[dat_combined$gene %in% sm2_markers]
length(markers)

stopifnot(all(markers %in% rownames(sig_sm2)))
stopifnot(all(markers %in% rownames(bulk_tpm)))

sig_sm2_tpm = get_tpm(sig_sm2, dat_combined$gene_length)
sig_10x_tpm = get_tpm(sig_10x, dat_combined$gene_length)

sig_sm2_tpm = sig_sm2_tpm[match(markers, rownames(sig_sm2_tpm)),]
sig_10x_tpm = sig_10x_tpm[match(markers, rownames(sig_10x_tpm)),]

bulk_tpm 	= bulk_tpm[match(markers, rownames(bulk_tpm)),]

dim(bulk_tpm)
dim(sig_sm2_tpm)

res_sm2 = ICeDT(bulk_tpm, sig_sm2_tpm, tumorPurity=rep(0,ncol(bulk_tpm)))
res_10x = ICeDT(bulk_tpm, sig_10x_tpm, tumorPurity=rep(0,ncol(bulk_tpm)))

rho_sm2 = t(res_sm2$rho)
rho_10x = t(res_10x$rho)

dim(rho_sm2)
rho_sm2[1:2,]

dim(rho_10x)
rho_10x[1:2,]

rhos = cbind(rho_sm2[,-1], rho_10x[,-1])
rhos = as.data.frame(rhos)
dim(rhos)

glist = list()
cols = brewer.pal(n = length(cell_types), name = "Dark2")

for(i in 1:length(cell_types)){
  cr_i = cor(rhos[[cell_types[i]]], 
              rhos[[cell_types_10x[i]]], 
              method="spearman")
  
  glist[[cell_types[i]]] = 
    ggplot(rhos, aes_string(x=cell_types[i], 
                            y=cell_types_10x[i])) + 
    geom_point(alpha=0.5, col=cols[i]) + 
    labs(title=sprintf("%s corr=%.3f", cell_types[i], cr_i), 
         x = "SMART-seq reference", 
         y = "10x reference")
}

ggarrange(plotlist=glist, nrow = 2, ncol = 3)

```


# Session information
```{r}
gc()

sessionInfo()
```

# References


